pyam_analysis/core.py

```diff
@@ -65,71 +65,94 @@ class IamDataFrame(object):
-        regions: list
-            list of regions to be imported
+        data: ixmp.TimeSeries, ixmp.Scenario, pandas.DataFrame or data file
+            an instance of an TimeSeries or Scenario (requires `ixmp`),
+            or pandas.DataFrame or data file with IAMC-format data columns
+        regions : list of strings
+            filter by regions
+        variables : list of strings
+            filter by variables (only with ixmp objects)
+        units : list of strings
+            filter by units (only with ixmp objects)
+        years : list of integers
+            filter by years (only with ixmp objects)
         """
-        # copy-constructor
-        if isinstance(data, IamDataFrame):
-            self.data = data.data
-            self._meta = data._meta
-            self.cat_color = data.cat_color
-            self.col_count = data.col_count
-
-        # read data from source
+        # import data from pandas.DataFrame or read from source
+        if isinstance(data, pd.DataFrame):
+            self.data = format_data(data)
+        elif has_ix and isinstance(data, ixmp.TimeSeries):
+            self.data = read_ix(data, regions=regions, variables=variables,
+                                units=units, years=years)
         else:
-            if has_ix and isinstance(data, ixmp.TimeSeries):
-                self.data = read_ix(data, regions)
-            else:
-                self.data = read_data(data, regions, **kwargs)
+            self.data = read_data(data, regions, **kwargs)
 
-            # define a dataframe for categorization and other meta-data
-            self._meta = return_index(self.data, mod_scen,
-                                      drop_duplicates=True)
-            self.reset_category(True)
+        # define a dataframe for categorization and other meta-data
+        self._meta = return_index(self.data, mod_scen,
+                                  drop_duplicates=True)
+        self.reset_category(True)
 
-            # define a dictionary for category-color mapping
-            self.cat_color = {'uncategorized': 'white', 'exclude': 'black'}
-            self.col_count = 0
+        # define a dictionary for category-color mapping
+        self.cat_color = {'uncategorized': 'white', 'exclude': 'black'}
+        self.col_count = 0
 
-    def append(self, other, regions=None):
-        """Read timeseries data and append to IamDataFrame
+    def append(self, other, inplace=False, regions=None, variables=None,
+               units=None, years=None, **kwargs):
+        """Import or read timeseries data and append to IamDataFrame
 
         Parameters
         ----------
-        data: ixmp.TimeSeries, ixmp.Scenario or data file
-            this option requires the 'ixmp' package as a dependency
-            or a file with IAMC-style data
-        regions: list
-            list of regions to be imported
+        other: pyam-analysis.IamDataFrame, ixmp.TimeSeries, ixmp.Scenario,
```

Comment:
it seems like there's a lot of conversion and stuff going on here. I would suggest sectioning that logic into two possible functions: `iam.read_data` (already exists) and `iam.to_dataframe` (which i think needs to be written. The first would read data from external sources, the second would translate in-memory data into an `iam.DataFrame`. The user would then be responsible for converting their data and providing it into these functions. The functions would therefore only take other iam.DataFrame objects.

Reply:
ah, looks like the `format_data` function does mostly this.

Reply:
I think that a function that takes many kinds of inputs is preferable from a user perspective, even if the docstrings are long. I think that internally, it's not quite as messy as it could have been...

Reply:
I don't think we disagree here. I'm arguing that there should be one or more "external functions" like `pd.read_csv` that support reading other datatypes and that all internal functions should assume (and check) that the type is an iam data frame. We can leave this for further cleanup if you wish.